### DISCUSSION TITLE
can we create type synonym

### DISCUSSION THREAD

**Discussion by wangyoutian** [dotnet/csharplang#4571] [Language Ideas]
can we create type synonym in C# like what we can do in SqlServer where we can create synonym for an database object such as table?

The syntax can be:

synonym A:n1.B{ // where A is the synonym of "n1.B"
}

Note here this is different from alias in that:
1) the synonym is defined in referenced lib, rather than referencing lib.
2) the synonym can used in anywhere, rather in a single file
3) no need for "using A=n1.B;" anymore.

**Comment by Sunlighter** [comment]
I&rsquo;d like something like this, too, but different. I think &ldquo;using&rdquo; should be extended in a couple of ways.

Currently `using` only applies to a file and does not appear in an assembly. A `using` does not belong to a namespace, either; only to a file.

Maybe there should be a `using class` syntax which *does* create a synonym in the assembly, where the synonym *does* belong to a namespace. That way, synonyms would show up in Intellisense and the like when you reference the assembly.

Calling it `using class` makes it unnecessary to define new keywords, and `using class` is nonsensical in existing C#, so introducing a `using class` syntax won&rsquo;t break any existing code. (But it&rsquo;s hard to Google it...)

The best implementation might be for the compiler to generate a trivial wrapper type with overloaded conversion operators.

Another alternative would be to define some new attribute in `System.Runtime.CompilerServices`, which could be applied to an assembly or a type and would mean that a synonym was defined inside. This would permit the compiler to erase the synonyms while compiling, but the synonyms would still be visible in the metadata, as attributes. The problem there is that existing reflection code would have to be modified to read the new attribute, whereas existing reflection code would already be able to see a trivial wrapper type. The trivial type would probably still have an attribute indicating that it was compiler-generated, though.

I&rsquo;m assuming we are generating synonyms for *types* only, and so this idea does not include synonyms for properties, methods, fields, events, constructors, etc.

`using class` should support generics, like `using class MyValues<K, V> = ImmutableSortedDictionary<K, MyKindOfValue<K, V>>`

`using class` should support `implicit` and `explicit`. I don&rsquo;t know which should be the default.

* The compiler regards an `implicit` synonym as the same type, so if you wrote `using implicit class MyInt = int;` it would *not* be an error to provide a `MyInt` to a function expecting an `int` or vice-versa. (This is useful because you could later change the code to `using implicit class MyInt = long;` and recompile and see what breaks. It is similar to the old `using` but it supports generics.)
* The compiler regards an `explicit` synonym as a distinct type, so I could write `using explicit class FileID = int;` and then it *would* be an error to pass an `int` to a function expecting a `FileID` or vice-versa, although an explicit cast would still be possible. This also means that `ImmutableList<FileID>` would be a distinct type from `ImmutableList<int>`.

Note that with `using implicit class U = T;`, if `T` implemented an interface then `U` would, too, because they are the same type. But with `using explicit class U = T;` then `U` would not implement any interfaces, so you might have to define `IComparer<U>` which does the explicit cast to `T`.

`using class` should allow defining synonyms as members of classes. I have a class `Unifier<K, G, V>` with a method `Unify` which returns `Tuple<V, ImmutableList<Proposition<K, V>>>` and it would be nice to be able to call that `Unifier<K, G, V>.UnifyResult` instead. (I know I can just use `var` but not when I need to create another variable of a different type such as `ImmutableList<Unifier<K, G, V>.UnifyResult>` and so forth. I could also define my own type; it&rsquo;s tedious but possible. But so is implementing `async` by hand. There are much worse cases than the one I&rsquo;m describing here, where the synonym would do a lot more good than it does here.)

Synonyms might have to be `public` or `private` or `protected`, etc., so you can write `using explicit public class...`

Synonyms that are not visible outside the assembly can possibly be erased by the compiler.

I don't know if there should be a `using struct` also. Maybe if `using class` generates a class, `using struct` should generate a struct. But it&rsquo;s possible that the generated type should *always* be a struct.

I don't think there needs to be a `using delegate` because you can just use `using class` or `using struct` for that.

(There is this issue of &ldquo;constructor forwarding&rdquo; where, if you can construct a `T` with certain arguments, and `U` is a synonym for `T`, you should also be able to construct a `U` with the same arguments, because, after all, `U` is a synonym. This could be handled by actually defining appropriate constructors on the generated types, but it could also be handled by some compiler trickery, where the generated type for `U` actually only has one constructor taking a `T`, and the compiler, seeing a `U` being constructed, would actually call the corresponding constructor for `T` and then pass the `T` to the constructor for `U`. Compiler trickery has the advantage that it would also work for delegates; if I recall, the compiler has special syntaxes for constructing delegates anyway, and there is no way to give those special syntaxes to a regular class or struct.)
